Map load unit testing and directory whitelisting - Mojave Sun is a really cool downstream, like and subscribe for longer PR names. oranges was here and made this github PR name much longer, thereby proving once and for all that upstreams do add value to the downstream not just endless amounts of refactoring because we can't write any api's in a modular fashion the first time around so have to update them later to be modular. Anyway do you think a taco counts as a sandwich? ~mothblocks wuz here

AndrewL97 opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->
This is a fix for map not load anymore because of security changes

## Why It's Good For The Game

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->
Maps are good as they encourage gameplay and differentiate ss13 from a classic MUD game

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl:
add: unit test for map load
add: directory param to map load + whitelist for data and _maps
add: advertising for mojave sun in tg commit logs
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

AndrewL97 committed:
I just want to make stupid game, and you guys keep breaking my downstream

JohnFulpWillard commented:
:trollface:

Mothblocks commented on code/modules/unit_tests/load_map_security.dm:
```diff
@@ -0,0 +1,26 @@
+//Tests to ensure we can load a map from a whitelisted directory (_maps), but not a non-whitelisted directory (i.e "fartyShitPants")
+/datum/unit_test/load_map_security/Run()
```
```suggestion
/// Tests to ensure we can load a map from a whitelisted directory (_maps), but not a non-whitelisted directory (i.e "fartyShitPants")
/datum/unit_test/load_map_security

/datum/unit_test/load_map_security/Run()
```

AndrewL97 replied:
Done

Mothblocks commented on code/modules/unit_tests/load_map_security.dm:
```diff
@@ -0,0 +1,26 @@
+/datum/unit_test/load_map_security/Run()
+
+	//Replace runtimestation with another valid map .json if not present
+	var/goodMap = "runtimestation"
```
Make this a define in this file instead of a var.

AndrewL97 replied:
Done - let me know if you want this in _defines/ somewhere - I based this off of other unit tests' use of defines

Mothblocks commented on code/modules/unit_tests/load_map_security.dm:
```diff
@@ -0,0 +1,26 @@
+	//Attempt to load our configs
+	var/datum/map_config/good_config = load_map_config("_maps",goodMap)
+	var/datum/map_config/bad_config = load_map_config("data/load_map_security_temp",goodMap)
+	var/datum/map_config/proove_fcopy_config = load_map_config("data","load_map_security_temp/[goodMap]")
```
"proove"? What does this var name mean?

AndrewL97 replied:
I can't spell "Prove"
This test is just to prove that the fcopy into data/ worked by loading it with the whitelisted directory.
I have renamed and clarified it, let me know if it is still muddy

Mothblocks commented on code/__DEFINES/maps.dm:
```diff
@@ -26,6 +26,9 @@ require only minor tweaks.
 
 #define SPACERUIN_MAP_EDGE_PAD 15
 
+//List of directories we can load map .jsons from
```
```suggestion
/// List of directories we can load map .jsons from
```

Mothblocks commented on code/modules/unit_tests/load_map_security.dm:
```diff
@@ -0,0 +1,26 @@
+	//Proove that fcopy worked for bad_config, and the .json is in the "bad directory"
+	TEST_ASSERT(!proove_fcopy_config.defaulted, "Failed to load: data/load_map_security_temp/[goodMap]")
+
+	fdel("data/load_map_security_temp/")
```
Run this on `Destroy()`, not just at the end of the test.

Otherwise if the test fails, we'll be leaving crud on from before.

AndrewL97 replied:
Done!

optimumtact commented:
Very interested in this PR going through, especially with a unit test .

Thanks

AndrewL97 committed:
Fixed commenting + made stub for unit test object
made targeted test map a define
clarified stupid variable names
fixed directory destroy

AndrewL97 committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into Mojave_Sun_Is_The_Best

# Conflicts:
#	code/__DEFINES/maps.dm
#	code/_compile_options.dm
#	code/datums/map_config.dm

AndrewL97 committed:
Fix param order in call to load_map_config

AndrewL97 commented:
@optimumtact @Mothblocks ready to re-review when you're free

AndrewL97 committed:
use PATH_TO_NEXT_MAP_JSON when we make next map json

optimumtact commented:
This PR title is nice, but it could be longer
